### PR TITLE
Allow functions within BashCompletionFunction to run unconditionally, limit their scope to a particular command

### PR DESCRIPTION
**Pull request by cpfair**
It's a pain to include all custom completion functions in a single bash function, and constantly check to make sure you're generating suggestions for the right command/s. So, this PR adds new `__custom_func` variants to handle both points. `__custom_func`'s behaviour is unchanged, so this should be backwards-compatible with existing handlers (excepting behaviour intended to run once on script load, which will now run on every completion).

The following is copy-pasted from the docs:

The `BashCompletionFunction` member can be used to define bash functions which generate completion suggestions. Within `BashCompletionFunction` you can define some of the following four bash functions, which should modify `COMPREPLY` as appropriate.

| Function | Runs always | Runs when no earlier handler produced completion results |
| --- | --- | --- |
| **Runs for Command** | `__custom_unconditional_command_func` | `__custom_command_func` |
| **Runs for Command & children** | `__custom_unconditional_func` | `__custom_func` |

If you define more than one `__custom*_func`, they will execute in the following order (latest takes precedence). If there is more than one instance of a certain function defined in a command hiearchy during completion, the most specific function will supplant less specific functions (which will _not_ execute at all).
- `__custom_unconditional_func`
- `__custom_unconditional_command_func`
- `__custom_func`
- `__custom_command_func`

### CODE REVIEW

**Comment by spf13** [comment]
@eparis since you guys use this feature extensively, can you give it a review?

File: bash_completions.go [hunk 1]
```diff
@@ -177,6 +185,8 @@ __handle_noun()
 
 __handle_command()
 {
+	# Clear handlers meant to be command-specific
+	unset -f __custom_command_func __custom_unconditional_command_func
```
Comment (eparis) [inline bash_completions.go]: whitespace nit.

**Comment by eparis** [comment]
I'm trying to understand how to actually use these?  Should we replace the example with something that uses the new _command_ versions?

**Comment by anthonyfok** [comment]
Hi @cpfair!  Thank you for your contribution!  Could you please follow up your pull request which enhances bash completion?
I guess you must have missed @eparis' code review and the subsequent request for more information from over a year ago, where he asked:

> I'm trying to understand how to actually use these? Should we replace the example with something that uses the new command versions?

Thank you!

**Comment by cpfair** [comment]
If I remember correctly, the issue was that the normal method of providing custom autocomplete was only useful when used on the root function - as the docs noted (see diff, where they were changed). This meant that if you wanted different autocomplete behavior for different command stems, you had to manually add & maintain that switching logic yourself. This change adds a built-in slot in which one can insert command-specific autocompletion functions, which are automatically selected based on the stem which the user has already typed. I'm honestly not sure what the unconditional variants were for.

In any case, if it's been this long and nobody else has raised a similar issue, it may not be worth the trouble of figuring out exactly what Past Me was thinking here...

**Comment by bogem** [comment]
@cpfair Will you solve all merge conflicts or should I close this PR?

**Pull request closed without merging**